### PR TITLE
*: use the revertable sandbox to buffer mutations

### PR DESCRIPTION
**Pull request by bobotu**
### What problem does this PR solve?
After #15780, this PR update the other parts of TiDB to use the new `Flush` and `Discard` API.

### What is changed and how it works?
Update the `MemBuffer` interface:

- Add `Flush` and `Discard` respect to `Sandbox`'s API
- Add `NewBuffer` to derive a new `Sandbox`
- Remove `Reset` and `SetCap` because there is no need to use them

The `TxnState`'s buffer now directly derived from the underlying `Transaction` when its state changed to `valid`. When statement commit/rollback the buffer will be discarded and a new `Sandbox` will be created (this allocation can be reduced by pooling temporal `Sandbox`, I'll submit a separate PR to add this optimization)

### CODE REVIEW

File: kv/memdb_buffer.go [hunk 1]
```diff
@@ -119,17 +115,33 @@ func (m *memDbBuffer) Delete(k Key) error {
...
 // Len returns the number of entries in the DB.
 func (m *memDbBuffer) Len() int {
-	return m.db.Len()
+	return m.sandbox.Len()
+}
+
+func (m *memDbBuffer) NewBuffer() MemBuffer {
+	return &memDbBuffer{
+		sandbox:         m.sandbox.Derive(),
+		entrySizeLimit:  TxnEntrySizeLimit,
```
[the remark this answers is not in the document]
Comment (bobotu) [reply kv/memdb_buffer.go:129]: 1. `entrySizeLimit` used to limit the size of `len(k) + len(v)`. The limit of this size should be the same across buffers?
2. Yes we should use `m.bufferSizeLimit - m.sandbox.Size()`, and this is the correct way to maintain the size limit.

**Comment by coocood** [comment]
LGTM

**Comment by tiancaiamao** [comment]
LGTM

File: kv/buffer_store.go [hunk 2]
```diff
@@ -37,24 +30,19 @@ type BufferStore struct {
...
 	}
 }
 
-// Reset resets s.MemBuffer.
-func (s *BufferStore) Reset() {
-	s.MemBuffer.Reset()
-}
-
-// SetCap sets the MemBuffer capability.
-func (s *BufferStore) SetCap(cap int) {
```
Comment (jackysp) [inline kv/buffer_store.go:56]: finally, it is removed.

**Pull request merged**